packages/lodestar/src/network/gossip/gossipsub.ts

```diff
@@ -0,0 +1,106 @@
+import Gossipsub, {IGossipMessage, Registrar, Options} from "libp2p-gossipsub";
+import {IGossipMessageValidator, LodestarGossipMessage} from "./interface";
+import {getGossipTopic, deserializeGossipMessage} from "./utils";
+import {GossipEvent} from "./constants";
+import {Attestation, AggregateAndProof, AttesterSlashing, ProposerSlashing, 
+  SignedVoluntaryExit, SignedBeaconBlock} from "@chainsafe/eth2.0-types";
+import {IBeaconConfig} from "@chainsafe/eth2.0-config";
+import {utils} from "libp2p-pubsub";
+import {ATTESTATION_SUBNET_COUNT} from "../../constants";
+
+/**
+ * This validates messages in Gossipsub and emit the transformed messages.
+ * We don't want to double deserialize messages for performance benefit.
+ */
+export class LodestarGossipsub extends Gossipsub {
+  private transformedMessages: Map<string, LodestarGossipMessage>;
+  private config: IBeaconConfig;
+  private validator: IGossipMessageValidator;
+
+  constructor (config: IBeaconConfig, validator: IGossipMessageValidator, peerInfo: PeerInfo,
+    registrar: Registrar, options: Options = {}) {
+    super(peerInfo, registrar, options);
+    this.transformedMessages = new Map();
+    this.config = config;
+    this.validator = validator;
+  }
+
+  public async validate(rawMessage: IGossipMessage): Promise<boolean> {
```

Comment:
This method is not really maintainable.
I would suggest to extract each validation to separate function, create validation map which maps event => validation function. Same for deserialization

Reply:
I added `getTopicValidator` and `getGossipObject` to make `validate` easier to read and maintain